### ISSUE TITLE
multi_head_attention_forward 3D attention mask incorrect behavior

### ISSUE THREAD

**Issue by Guitaricet** [pytorch/pytorch#63271]
## 🐛 Bug

After PyTorch 1.9, 3D masks in `multi_head_attention_forward`, when used with `key_padding_mask` cause `nan` values in the attention output.

## To Reproduce

1. Call `multi_head_attention_forward` with a 3D attention mask and a non-zero padding mask

## Expected behavior

The output of `multi_head_attention_forward` should not contain nan for a non-nan input.

## Additional context

The problem seem to arise in commit 90f05c005ddbd6eeda6cb387124afc07e4b46636

In particular the operation at the line https://github.com/pytorch/pytorch/blob/90f05c005ddbd6eeda6cb387124afc07e4b46636/torch/nn/functional.py#L5046

is not equivalent to the one before this commit. In the case of 3D masks and non-zero padding, new implementation creates 1s for both rows and columns of padded elements.
After filling them with `-inf` and softmaxing, this creates nan elements (because `softmax([-inf, -inf, ..., -inf])` is nan.

Example of incorrect mask (single head, dimension semantics [src_len, tgt_len], last sequence element is padded):

```
tensor([[0., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1.],
        [0., 0., 0., 1., 1.],
        [0., 0., 0., 0., 1.],
        [1., 1., 1., 1., 1.]])
```

Instead, it should be (as the code before the commit 90f05c005ddbd6eeda6cb387124afc07e4b46636 would effectively do)

```
tensor([[0., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1.],
        [0., 0., 0., 1., 1.],
        [0., 0., 0., 0., 1.],
        [0., 0., 0., 0., 1.],
```

Given this mask, none of the source elements can look at the padding (as required) and we don't feed `-inf`-filled vectors to softmax.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @bhosmer @cpuhrsch

**Comment by VitalyFedyunin** [comment]
Hi-Pri as it looks like regression

**Comment by bhosmer** [comment]
@Guitaricet hi Vlad - would you be able to give me the example inputs you're using to repro the difference between pre- and post-https://github.com/pytorch/pytorch/commit/90f05c005ddbd6eeda6cb387124afc07e4b46636 behavior? I've tried a few obvious examples and they all show the nans you're describing, but they happen *both* before and after that commit - I haven't found a repro that shows a change. If you have one, that would be super helpful. Thanks!

**Comment by Guitaricet** [comment]
I am getting confused too.

For the attention mask

```
      [[[False,  True,  True,  True,  True],
         [False, False,  True,  True,  True],
         [False, False, False,  True,  True],
         [False, False, False, False,  True],
         [False, False, False, False, False]]]
```

and padding mask

```
       [[False, False, False,  True,  True]]
```

and some input of shape `[5, 3, hidden]`

Multihead attention gives a nan-free output (for both 1.8 and 1.9).

Then I noticed that I forgot to transpose the attention mask, because multihead attention expects the order `[batch * heads, target, soruce]`.

After transposition I get nans for both 1.8 and 1.9.

Now I'm thinking maybe I have confused the dimension semantic for attention mask and you don't need to transpose the mask from `triu`.

**Comment by bhosmer** [comment]
Actually I think the attention mask in your example is already in the right form. Each row represents attention from a target (query) to the sequence of sources, so the lower triangular mask makes sense in terms of causality: the top row (first target) can only attend to the first column (first source), the second target to the first and second sources, and so on down to a completely unmasked final target. 

OTOH if you transpose this, it only allows attention from a target to present and *future* sources, which completely masks the final row, except for the final column - which was then being masked out by the key padding mask. 

If this analysis is correct, I think the takeaway is 
* while it is possible to supply a combination of attention mask and key padding mask that will produce NaNs, it's not an attention mask that you'd ever supply intentionally in practice.
* still, it's easier than it should be to supply such a mask accidentally - in particular, shapes won't help us catch mistranspositions when attention weights are square. But I'm not sure there's a good way to trap an error of this kind. Probably the right solution is a place in the upstream API that lets you just say you want causal attention and generates the mask for you. (cc @jbschlosser per our API discussions - also we should check the usual suspects for precedent :)

@Guitaricet Vlad, please LMK what you think of this conclusion - I'll leave the issue open until you've had a chance to take a look.
